### ISSUE TITLE
strict=False doesn't propagate as expected

### ISSUE THREAD

**Issue by kstrauser** [schematics/schematics#316]
When one model is nested inside another, importing that model doesn't pass `strict` through as expected. For example:

```python
    from schematics.models import Model
    from schematics.types import StringType
    from schematics.types.compound import ModelType

    class Inner(Model):
        foo = StringType()

    class Outer(Model):
        inner = ModelType(Inner)

    def main():
        d = {
            'inner': {
                'foo': 'foo value',
                'bar': 'bar value',
            },
        }

        outer = Outer(d, strict=False)
        print outer
        print outer.inner
```

This fails with:

```python
    Traceback (most recent call last):
      File "/private/tmp/schem/bin/main", line 9, in <module>
        load_entry_point('main==0.1.0', 'console_scripts', 'main')()
      File "/private/tmp/schem/schematics_rogue_field_issue/src/main/main.py", line 19, in main
        outer = Outer(d, strict=False)
      File "/private/tmp/schematics/schematics/models.py", line 240, in __init__
        self._data = self.convert(raw_data, strict=strict, mapping=deserialize_mapping)
      File "/private/tmp/schematics/schematics/models.py", line 287, in convert
        return convert(self.__class__, raw_data, **kw)
      File "/private/tmp/schematics/schematics/transforms.py", line 419, in convert
        partial=partial, strict=strict, mapping=mapping)
      File "/private/tmp/schematics/schematics/transforms.py", line 119, in import_loop
        raise ModelConversionError(errors, data)
    schematics.exceptions.ModelConversionError: {'inner': {'bar': 'Rogue field'}}
```

Instead, I would have expected it to successfully instantiate the `outer` object.

**Comment by kracekumar** [comment]
```python
In [20]: class Inner(Model):
    foo = StringType()
   ....:

In [21]: class Outer(Model):
    inner = ModelType(Inner, strict=False)
   ....:

In [22]: d = {'inner': {
'foo': 'v',
'bar': 'bad'}}

In [23]: v = Outer(d, strict=False)

In [24]: v.inner
Out[24]: <Inner: Inner object>

In [25]: v.to_primitive()
Out[25]: {'inner': {'foo': u'v'}}

In [26]: v.inner.to_primitive()
Out[26]: {'foo': u'v'}
```

Right way to do is to pass `strict=False` during declaration. 

Setting at model declaration step allows granular control.

**Comment by kstrauser** [comment]
I disagree. That's *way* too early in the data flow for me to know whether I need to use something so ephemeral. A common case is that I want strict parse by default, but one just is just not capable of getting their act together and not adding extra keys to a document, so I want to special-case their importer to use non-strict parsing. I'm don't want to lock down `strict=True` **or** `strict=False` at compilation time because at that point I don't have enough context to know which one I'll want. I certainly don't want to have `StrictOuter` and `LooseOuter` classes that I have to keep track of.

But more than that, it's surprising. When I write `outer = Outer(d, strict=False)`, I expect that it will use non-strict importing. That's not the case.

I'll code up a patch this morning if I can.

**Comment by kstrauser** [comment]
This is fixed in pull request #318.

**Comment by bintoro** [comment]
This is a special case of a larger issue. `strict` is not the only import option that fails to be propagated through multiple levels of recursion. Unfortunately, passing all these values along individually gets very messy because they all must appear in this huge chain of function calls.

Sometimes the argument list is even contingent on circumstances:
```python
try:
    ...
    raw_value = field_converter(field, raw_value, mapping=model_mapping)
except Exception:
    raw_value = field_converter(field, raw_value)
```

This is ridiculous.

Instead, I would like there to be a single metadata container that would hold `strict` and similar options. Adding further parameters would become a breeze – no need to modify a dozen functions.

I have tentatively developed this idea in PR #266. @kstrauser, can you please have a look?

**Comment by kstrauser** [comment]
I actually like that implementation a lot better. However, we already pass a `context` object down through the call stack. The current expectation is that a user passes it in, but could we leverage that anyway (copy the passed-in object; insert "_mapping" and "_force_strict" keys, etc.)?

**Comment by kstrauser** [comment]
As a side note, what do @kracekumar, @bintoro, and @jfitzpatrick99 think of the docs and test patches, particularly the semantics? In a nutshell, calling `MyModel(strict=<value>)`:

- Behaves exactly as normal if `None` (the default) is given
- Forces strict importing if `True`
- Forces unstrict importing if `False`

Implementation aside, does this look OK to you?

**Comment by jfitzpatrick99** [comment]
This looks good to me.

**Comment by bintoro** [comment]
> However, we already pass a `context` object down through the call stack. The current expectation is that a user passes it in, but could we leverage that anyway (copy the passed-in object; insert "_mapping" and "_force_strict" keys, etc.)?

Potentially yes, but so far I've been hesitant to go this route, as I can see a couple of drawbacks and not any major benefits.

In the scheme I developed in #266, `strict` etc. would continue to be available as arguments to `Model.__init__`, for instance. Whenever `env` is present, however, its contents will override the other arguments, which allows the parent loop to lock in their values. Now, if the environment were instead provided by `context`, then we'd have to be careful about where `context` might be entering the loop because it will override or reset the environment from that point on. This may result in unexpected behavior with some existing programs.

From a logical standpoint, as well, it would be ideal to store `context` inside the `env` object and not the other way around.

(As an aside, the whole `context` business may be in need of review. I'm actually a bit confused whether `context` is even intended for export operations currently.)

Finally, it seems that implementing the environment object is a breaking change mainly regarding custom types, and this occurs whether we reuse an existing parameter or not.

> In a nutshell, calling `MyModel(strict=<value>)`:
>
>   * Behaves exactly as normal if `None` (the default) is given
>   * Forces strict importing if `True`
>   * Forces unstrict importing if `False`

Yeah, this is how I've always thought import/export options should work.

**Comment by kstrauser** [comment]
Packing `context` inside `env` makes a lot of sense. I like that.

I'd explicitly intended `context` for export (although I could imagine it having import uses, too). The example from the docs mentioned localization, where you might have one instance that holds multiple export possibilities and the context allows you to switch between them at runtime.

**Comment by kracekumar** [comment]
I am with @bintoro on this.

**Comment by bintoro** [comment]
Fixed in #359.
